cura/OAuth2/AuthorizationService.py

```diff
@@ -0,0 +1,162 @@
+
+    def getAccessToken(self) -> Optional[str]:
+        """
+        Get the access token response data.
+        """
+        if not self.getUserProfile():
+            # We check if we can get the user profile.
+            # If we can't get it, that means the access token (JWT) was invalid or expired.
+            return None
+
+        if self._auth_data is None:
+            return None
+
+        return self._auth_data.access_token
+
+    def refreshAccessToken(self) -> None:
+        """
+        Refresh the access token when it expired.
+        """
+        if self._auth_data is None or self._auth_data.refresh_token is None:
+            Logger.log("w", "Unable to refresh access token, since there is no refresh token.")
+            return
+        self._storeAuthData(self._auth_helpers.getAccessTokenUsingRefreshToken(self._auth_data.refresh_token))
+        self.onAuthStateChanged.emit(logged_in=True)
+
+    def deleteAuthData(self) -> None:
+        """Delete authentication data from preferences and locally."""
+        self._storeAuthData()
+        self.onAuthStateChanged.emit(logged_in=False)
+
+    def startAuthorizationFlow(self) -> None:
+        """Start a new OAuth2 authorization flow."""
+
+        Logger.log("d", "Starting new OAuth2 flow...")
+
+        # Create the tokens needed for the code challenge (PKCE) extension for OAuth2.
+        # This is needed because the CuraDrivePlugin is a untrusted (open source) client.
+        # More details can be found at https://tools.ietf.org/html/rfc7636.
+        verification_code = self._auth_helpers.generateVerificationCode()
+        challenge_code = self._auth_helpers.generateVerificationCodeChallenge(verification_code)
+
+        # Create the query string needed for the OAuth2 flow.
+        query_string = urlencode({
+            "client_id": self._settings.CLIENT_ID,
+            "redirect_uri": self._settings.CALLBACK_URL,
+            "scope": self._settings.CLIENT_SCOPES,
+            "response_type": "code",
+            "state": "CuraDriveIsAwesome",
+            "code_challenge": challenge_code,
+            "code_challenge_method": "S512"
+        })
+
+        # Open the authorization page in a new browser window.
+        webbrowser.open_new("{}?{}".format(self._auth_url, query_string))
+
+        # Start a local web server to receive the callback URL on.
+        self._server.start(verification_code)
+
+    def _onAuthStateChanged(self, auth_response: AuthenticationResponse) -> None:
+        """Callback method for an authentication flow."""
+        if auth_response.success:
+            self._storeAuthData(auth_response)
+            self.onAuthStateChanged.emit(logged_in=True)
+        else:
+            self.onAuthenticationError.emit(logged_in=False, error_message=auth_response.err_message)
+        self._server.stop()  # Stop the web server at all times.
+
+    def _loadAuthData(self) -> None:
```

Comment:
I think this should be in the `Account` class. This service should just do things like launching a browser for login and handling login callbacks. It also makes testing easier.

Reply:
Do you mean make it public (and let the account class call the loading of the data or fully move to the account class?)